scheduler/src/cook/kubernetes/api.clj

```diff
@@ -1323,263 +1187,367 @@
+        ; (https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#node-affinity)
+        (let [affinity (V1Affinity.)
+              node-affinity (V1NodeAffinity.)
+              node-selector (V1NodeSelector.)
+              node-selector-term (V1NodeSelectorTerm.)]
+
+          ; Disallow scheduling on hostnames we're being told to avoid (if any)
+          (let [node-selector-requirement-k8s-hostname-label (V1NodeSelectorRequirement.)]
+            (.setKey node-selector-requirement-k8s-hostname-label k8s-hostname-label)
+            (.setOperator node-selector-requirement-k8s-hostname-label "NotIn")
+            (run! #(.addValuesItem node-selector-requirement-k8s-hostname-label %) pod-hostnames-to-avoid)
+            (.addMatchExpressionsItem node-selector-term node-selector-requirement-k8s-hostname-label))
+
+          (.addNodeSelectorTermsItem node-selector node-selector-term)
+          (.setRequiredDuringSchedulingIgnoredDuringExecution node-affinity node-selector)
+          (.setNodeAffinity affinity node-affinity)
+          (.setAffinity pod-spec affinity))))
+
+    (.setRestartPolicy pod-spec "Never")
+
+    (.addTolerationsItem pod-spec toleration-for-deletion-candidate-of-autoscaler)
+    (.addTolerationsItem pod-spec (toleration-for-pool cook-pool-taint-name cook-pool-taint-prefix pool-name))
+    ; We need to make sure synthetic pods --- which don't have a hostname set --- have a node selector
+    ; to run only in nodes labelled with the appropriate cook pool
+    (when-not hostname (add-node-selector pod-spec cook-pool-label-name pool-name))
+
+    (when pod-constraints
+      (doseq [{:keys [constraint/attribute
+                      constraint/operator
+                      constraint/pattern]}
+              pod-constraints]
+        (condp = operator
+          :constraint.operator/equals
+          ; Add a node selector for any EQUALS constraints
+          ; either defined by the user on their job spec
+          ; or defaulted on the Cook pool via configuration
+          (add-node-selector pod-spec attribute pattern)
+          :else
+          ; A bad constraint operator should have
+          ; been blocked at job submission time
+          (throw (ex-info "Encountered unexpected constraint operator"
+                          {:operator operator
+                           :task-id task-id})))))
+
+    (.setVolumes pod-spec (filterv some? (conj volumes
+                                               init-container-workdir-volume
+                                               scratch-space-volume
+                                               sidecar-workdir-volume
+                                               checkpoint-volume)))
+    (.setSecurityContext pod-spec security-context)
+    (.setPriorityClassName pod-spec pod-priority-class)
+
+    ; pod
+    (.setMetadata pod metadata)
+    (.setSpec pod pod-spec)
+
+    (when (synthetic-pod? pod-name)
```

Comment:
And here we do do a `synthetic-pod?`... Why?

Reply:
See above.